<h1 style="text-align:center">Classificação de vinho</h1>
<h4>Objetivo:</h4>
<p>Utilizar aprendizagem de maquina baseado em arvore de classificação para corelacionar caracteristicas dos vinhos com o tipo de vinho sendo, tinto ou branco</p>
<img style="text-align:center" src="vinho_capa.jpg">

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import ExtraTreesClassifier

In [2]:
df = pd.read_csv('wine_dataset.csv')

In [3]:
df.head()

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality style  
0      9.4        5   red  
1      9.8        5   red  
2      9.8        5   red  
3      9.8        6   red  
4      9.4        5   red

<h5>Pequeno ajuste alterando o nome das variaveis de tinto para 1 e branco para 0</h5>

In [6]:
df['style'].replace('red',1,inplace=True)
df['style'].replace('white',0,inplace=True)
df.head()

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  style  
0      9.4        5      1  
1      9.8        5      1  
2      9.8        5      1  
3      9.8        6      1  
4      9.4        5      1

<h5>Marcamos nossa variavel Target no qual desejamos prever do dataset original  e criamos um  DF sem a target na qual  iremos prever</h5>

In [7]:
#separamos o df original em Variavel Target e variavel preditiva
target = df['style']
predict = df.drop('style',axis=1)

<h4>Extra Tress Classifier:</h4>
<p>iremos utilizar o modelo de arvores extras de classificação:<br>Extra Trees é como a Random Forest, na medida em que constrói várias árvores e divide nós usando subconjuntos aleatórios de recursos, mas com duas diferenças principais: ele não inicializa observações (o que significa amostras sem substituição) e os nós são divididos em divisões aleatórias, não melhores divisões. Então, em resumo, ExtraTrees:</p>
<img src="arvore.png">

<p>Nosso dataset será separado em 75% para treino e 25% para teste</p>

In [8]:
predict_trainning,predict_test,target_trainning,target_test = train_test_split(predict,target,test_size = 0.25)
#Utilizamos arvore de descissão
model = ExtraTreesClassifier()
#função fit espera como parametro o df e o array
model.fit(predict_trainning,target_trainning)
#pontuação que o modelo obteve de acerto
result = model.score(predict_trainning,target_trainning)
print('Resultado de acurácia:',result)

Resultado de acurácia: 0.9995894909688013


c:\users\drt903126\appdata\local\programs\python\python36-32\lib\site-packages\sklearn\ensemble\forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


<h5>Tivemos um resultado de acurácia de 99%<h5>

In [9]:
#Realizamos a predição com um df que o modelo ainda não conhece
predict_teste = model.predict(predict_test)

In [10]:
#criando uma coluna novamente no df que retiramos e adicionamos o resultado do array já com as "respostas"
predict_test['type'] = predict_teste
#voltamos aos nomes originais
predict_test['type'].replace(1,'red',inplace=True)
predict_test['type'].replace(0,'white',inplace=True)
#Resultado:
predict_test.head()

c:\users\drt903126\appdata\local\programs\python\python36-32\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
c:\users\drt903126\appdata\local\programs\python\python36-32\lib\site-packages\pandas\core\generic.py:6586: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._update_inplace(new_data)


fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
1614            6.6             0.170         0.38             1.5      0.032   
546             7.5             0.550         0.24             2.0      0.078   
2547            6.8             0.705         0.25             3.2      0.048   
4584            6.7             0.180         0.28            10.2      0.039   
4334            6.5             0.350         0.28            12.4      0.051   

      free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
1614                 28.0                 112.0  0.99140  3.25       0.55   
546                  10.0                  28.0  0.99830  3.45       0.78   
2547                 10.0                  57.0  0.99600  3.36       0.52   
4584                 29.0                 115.0  0.99469  3.11       0.45   
4334                 86.0                 213.0  0.99620  3.16       0.51   

      alcohol  quality   type  
1614     11.4        7  white  
546       9.5        6    red  
2547      9.5        4    red  
4584     10.9        7  white  
4334      9.9        6  white

<h3>Resultado</h3>
<p>no final tivemos 1256 vinhos brancos e 369 vinhos tintos segundo nosso modelo.</p>

In [34]:
predict_test['type'].value_counts()

white    1256
red       369
Name: type, dtype: int64

<h4>E como esse modelo se comporta com outro dataset?</h4>

In [23]:
#Mas será que o modelo realmente treinou? Testaremos com outro dataset obtido no kaggle.
df_teste = pd.read_csv('winequalityred_test.csv')

<p>Teremos as mesmas variaveis que tinha nosso modelo que utilizamos para treino porem dessa vez é um modelo apenas de vinho tinto</p>

In [24]:
df_teste.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

<p>Realizamos a chamada para predição</p>

In [25]:
predict_teste_2 = model.predict(df_teste)

In [27]:
df_final = df_teste.copy()

In [32]:
#criando uma coluna novamente no df que retiramos e adicionamos o resultado do array já com as "respostas"
df_final['type'] = predict_teste_2
#voltamos aos nomes originais
df_final['type'].replace(1,'red',inplace=True)
df_final['type'].replace(0,'white',inplace=True)


<h4>Resultado</h4>
<p>Será que nosso modelo teve foi acertivo na sua predição?</p>

In [33]:
df_final['type'].value_counts()

red      1591
white       8
Name: type, dtype: int64

<p>Das 1599 linhas que continha nosso dataset o mesmo conseguiu classificar como 1591 vinhos vermelhos e apenas 8 como branco uma taxa pequena se for considerar que nosso dataset por inteiro é apenas de vinho tinto</p>